### ISSUE TITLE
What is the point of the no-continue rule?

### ISSUE THREAD

**Issue by callumlocke** [airbnb/javascript#1103]
The official ESLint docs [say](http://eslint.org/docs/rules/no-continue) "When used incorrectly it makes code less testable, less readable and less maintainable."

Does Airbnb agree with the above reasoning? If so, can you give an example of how `continue` makes code less testable and less maintainable?

(I'm not arguing against it btw, just want to understand why.)

**Comment by ljharb** [comment]
continue, like break, is GOTO. There are countless articles on the internet about GOTO being harmful.

In addition, this style guide discourages loops, which are the only places you'd need continue or break.

**Comment by deckar01** [comment]
`continue` is generally not used with a label. I think eslint needs to differentiate GOTO label usage from bare `continue` usage. When used without a label, it is no different than a `return` statement in the body of a `forEach` loop.

I am going to disable this rule for my project. Would you be interested in added granularity for the `no-continue` rule if I proposed it?

**Comment by deckar01** [comment]
https://github.com/airbnb/javascript/blob/90235cab7c060d69430001cf17aa28ff74da86b4/packages/eslint-config-airbnb-base/rules/best-practices.js#L128

[`no-labels`](https://eslint.org/docs/rules/no-labels) is already enabled.

**Comment by ljharb** [comment]
Loops themselves are banned by this gude; so there’s no use case for continue.

**Comment by deckar01** [comment]
I am using the latest version of the airbnb base config for eslint and it doesn't warn about `for` loops.

**Comment by deckar01** [comment]
Also, how would you generate an array with the numbers A..B without using a `for` loop? My guess is that it is less readable than

```js
const range = [];
for (let x = A; x <= B; x += 1) {
  range.push(x);
}
```

**Comment by ljharb** [comment]
The guide is far more expansive than the linter config; rest assured, it will one day warn on for loops, to match the guide text.

The answer is you’d make a “range” abstraction, in which the implementation details are both irrelevant and fine to override linter warnings on, rather than writing a loop in your own code.

**Comment by SelaO** [comment]
Continue is not goto. That's like saying loops are goto.

**Comment by ljharb** [comment]
@SelaO no, it's not like saying that. Loops that use `break` or `continue` are goto.

**Comment by SelaO** [comment]
What's the difference between the following?

```
start:
...
if(x) goto start
```

and

```
start:
...
if(y) goto start
...
if(x) goto start
```

and

```
start:
...
if(y) goto end
...
if(x) goto start
end:
```

Now how are these examples (which look like nested loops and keep all the jumps inside the scope of the loop) are the same as:

```
function foo(){
...
ARBITRARYLABEL1:
....
goto ARBITRARYLABEL2
}

function bar(){
...
goto ARBITRARYLABEL1
...
}
```

**Comment by MegaArman** [comment]
I actually don't currently use airbnbs rules, but am considering it after reading the discussion here.
 continue should be banned, especially if you don't use loops (you shouldn't in 2018, but that's a whole different discussion).

I was once a continue, break, return everywhere, kind of dude and felt very clever for leveraging those features, but really it doesn't help your team. The less bouncing around you have to do when reading code the better. continue makes code harder to read for other people -you might not see why from your own code, but imagine working on a large codebase with lots of loops and if statements... Do you really want to maintain code you can't read top to bottom in one shot? Bite the bullet, learn to write easier to read code even if it's initially harder to do.

Just for reference, in Scala the last line of code is automatically the return value if no return is specified and it's what they seem to encourage (disclaimer: I'm not a Scala expert, but just saying that there's things to be learned from other languages).

**Comment by SelaO** [comment]
Loops are still useful in 2018 in many non trivial cases and will stay useful even in 2118, you can't do everything with map, reduce, filter.

See the example I gave above and tell me how continue is "bouncing around" like goto is. 

> but imagine working on a large codebase with lots of loops and if statements with continues scattered about

That's an invalid statement, if the code already become such a mess, then not having continue and breaks won't help you, loops shouldn't be huge, just like functions shouldn't be huge. 
Banning basic coding concepts isn't a magic tool to make code more readable and maintainable, and oftenly can make the code even less maintainable (like more complex stopping conditions in no continue loops)

**Comment by ericblade** [comment]
Yes, there are places where you should use loops.  If you're doing it in front-end code, inside your main code-base, though, it might be a sign that there are other things that could be problems.  In general, however, your arrays should be things that you want to iterate completely over.  You may have a data structure problem, if you find yourself needing to drop out of a render loop early, for example.

**Comment by ericblade** [comment]
There are places where you can make a convincing argument for just about any language construct, even the ones that are almost universally panned.  Ask a kernel programmer about goto, for example. 

Here's an easy example where a for loop is useful.  Say you have 5000 records in an array, but you want the first 200 that fit a particular criteria.  So you write a loop to search that array, and break once the limit is hit.  Perfect use for a for loop.  But you should abstract that to a separate function outside of your main logic.  If you're doing that in a front-end environment, I'd also seriously consider if the front-end is the place to do that processing -- do you need all that data on the front-end, or should you be sending only the first 200 that fit from the back-end to begin with?  That depends on your use case.  Either way, it shouldn't be in the main logic.

**Comment by ljharb** [comment]
Or you write an abstraction using .some or .find, no loop required. Totally agree with the rest of the last comment tho.

**Comment by ericblade** [comment]
.... .some or .find return a true/false or a single item . . i could see ways to use them to construct something, then break, but i feel like that would be abusing them in a way that doesn't make sense, for the sake of not using a traditional loop.

**Comment by ljharb** [comment]
It’s not abusing them, it’s using them - iteration is best expressed with iteration methods, not with loops.

**Comment by ericblade** [comment]
I'd like to see an example of using .some or .find to return a limit 200 of a larger array, that doesn't look like abusing them, if you've got a minute to teach me. :-)

**Comment by ljharb** [comment]
```js
function take(arr, predicate, count = Infinity) {
  const taken = [];
  arr.some((item, ...args) => {
    if (predicate(item, ...args)) {
      taken.push(item);
    }
    return taken.length >= count;
  });
  return taken;
}
```

Obviously you may consider that “abuse” - that’s subjective - i do not; i consider that a perfectly reasonable usage of `some` when hidden by a presumably well-tested abstraction.

**Comment by Zakini** [comment]
> Loops themselves are banned by this gude

I feel like I've seen this repeated far too many times across multiple issues. Would it be worth adding `'ForStatement'` and `'WhileStatement'` to `no-restricted-syntax`? That way rules related to loop-only syntax (are there any other than `no-continue`?) could be safely removed. If anyone decides to disable `no-restricted-syntax` in their project, they can use loops in any way they want without having to disable extra rules; they're off the style guide by that point anyway.

That said, discussion around banning loops always revolves around `for`, `for-of`, etc. and `while` never gets mentioned. Is `while` actually allowed by this guide? I can't think what the functional alternative to it would be...

**Comment by ljharb** [comment]
Eventually we’ll be adding those rules, but we’d ever want to remove no-continue or anything similar; even when using loops, using continue is GOTO.

**Comment by Venryx** [comment]
> SelaO: Continue is not goto. That's like saying loops are goto.

> ljharb: no, it's not like saying that. Loops that use break or continue are goto.

Lets look at an example.

Using continue:
```
for (let item of array) {
  if (item == 0) {
    continue;
  }
  console.log("Found non-zero item");
}
```

Using if:
```
for (let item of array) {
  if (item != 0) {
    console.log("Found non-zero item");
  }
}
```

I'm guessing you'd say that the continue example is like GOTO because: **it snaps the point of execution** (from line-3 in example 1 to line-2, for the next iteration)

The thing is, "snapping the point of execution" is also happening in example 2. You're "snapping the point of execution" from line 2 to line 4.

"Okay, but in the second example, your always snap in the same direction -- you always snap downward, to directly after the block's braces."

The thing is, you can make a similar statement about continue: "Okay, but you always snap in the same direction -- you snap upward, to the first line of the loop."

In other words, both continue statements and if statements snap execution, and they do so in a consistent way: one always snaps up, the other always snaps down.

Conclusion: I don't see how the continue statement is significantly more "like a GOTO" than an if statement.
= = = = = = = = = =
The main argument I can see in response is this: "Both statements act consistently, however continue statements obscure readability more because sometimes the start of the loop is on a much higher nesting level than the continue statement, requiring more traversal to find the next executed line."

I think that's a fair point, however:
1) Many (if not most) times, usages of continue are done in simple if statements directly under the for loop, in which case you don't have to traverse through nesting levels -- so it's equivalent to an if-statement in traversal distance.
2) Even in the nested cases where continue statements may yield worse readability (depends on how you'd have to restructure to avoid the continue), it's still an exaggeration to say "continue statements are GOTO" whereas "if statements are nothing like GOTO".

Both of them:
* Snap the point of execution.
* But they do so consistently.
* With the programmer knowing unambiguously where the next line of execution will be. (without having to search for labels like with GOTO)
* And with no chance of jumping to areas of code which have code prior to them that (misleadingly) have not yet been run. (which is one of the main problems with GOTO)

I'd like to see a list of properties of continue statements which make them equivalent to GOTO statements, which are not shared by loops or if statements. I know there are some, but I don't believe them to be substantial enough to say "continue is GOTO" but "if statements are nothing like GOTO".

**Comment by ericblade** [comment]
For me, I ignore the rules that encourage multiple return statements, and I stick to a rule that I've held much longer: There should be at most (with some potential exceptions) two points of exit to a function -- throw at the top, return at the bottom.  Of course, we can't always throw when we need to exit early, but in general: early exit as quick as possible at the top, otherwise flow should reach the bottom.

Your opinions may differ, and of course, this is the AirBnB style guide, not ericblade's style guide.  but if i were to write one, it would be quite similar to this one :-)

**Comment by Venryx** [comment]
Yep, and I can understand that line of thinking. My intention's not to change the airbnb style, and I normally don't comment so long on these things. The main reason I responded was because of the insistence that kept being made that "continue is GOTO".

In my opinion, that designation is lazy and misleading, because several of the negative properties of GOTOs are not in fact shared by continues. And in fact, the properties that continue does share with GOTO are all shared by other accepted constructs.

So basically, the "continue is GOTO" rhetoric is what got me riled up (like a false assertion at a political rally), and if that one point were retracted I'd be content.

**Comment by mikkqu** [comment]
Totally agree. Advocating against `continue` in functions is pretty much the same as advocating against `return guard clauses` in normal functions.

**Comment by hcbh96** [comment]
Let say I am trying to stick to the rule of no GOTO as suggested and I have a piece of code as follows what can I use to make the code more readable? 

- the code must print the Lowest number of matching objects in each array i.e for the two supplied it should print 1 X 1s & 2 X 2s

```
const array1 = [1, 2, 2, 2, 2, 2, 'h', 'h', 'h', '', 'b', 0, 'r', 10, 'e', 'a', 12, 'k', '', 99, 'i', 's', 100, '', 'g', 101, 'r', 'e', 'a', 't'];
const array2 = [1, 2, 5, 1, 2, 'h', 'h', 'h', 'h', '', 't', 'a', 1003, 'e', 'r', 'g', 1012, '', 15, 's', 'i', '', 999, 'k', 'a', 200, 'e', 'r', 1001, 'b'];
let log = 1;
for (let i = 0 - 1; i < array1.length; i += 1) {
  const _1 = array1[i];
  // eslint-disable-next-line no-labels
  for (let j = array2.length - 1; j >= 0; j -= 1) {
    const _2 = array2[j];
    // compare length first may save time
    if (_2 === _1) {
    // contacts have identical details update server contact and remove from array
      console.log(_2);
      log += 1;
      array2.splice(j, 1);
      break;
    }
  }
}
```

**Comment by dcbrwn** [comment]
Deep down inside, all the control flow statements are goto's. Saying that break and continue are goto's, is not enough to justify the rule, because we can say the same for `if`s and loops. Even function calls are fancy `goto`s.

**Comment by ljharb** [comment]
What things are “deep down” isn’t relevant; deep down, a gasoline-based car is a bunch of explosions, but a car isn’t an explosion.

**Comment by dcbrwn** [comment]
That's absolutely true. Therefore saying that we shouldn't use cars because they are moving by exploding - is not a valid statement.

**Comment by SelaO** [comment]
@dcbrwn you're wasting your time. Airbnb isn't the god of js.
Just edit that rule and move on.

**Comment by ljharb** [comment]
@dcbrwn sure. the general rule would be "don't use explosions", just like it's "don't use goto" - but things that are "basically goto" but aren't *actually* goto aren't included in that prohibition.

**Comment by Venryx** [comment]
> but things that are "basically goto" but aren't actually goto aren't included in that prohibition.

Exactly. Which is why break and continue should be allowed, just as if statements, return statements, function calls, etc. are.

In other words, it has yet to be shown how break and continue are "more like goto" than those other elements above. The reasons so far provided have been discussed at length above, and shown to be present in those other allowed elements, so they are not valid as distinguishers between break/continue and those other elements.

To format that more explicitly, it's being claimed that:
1) break and continue "are goto"
2) if statements, return statements, loops, and function calls are not goto

But when challenged to say how break and continue are "more like goto", the reasons provided do not stand up to scrutiny because those reasons/properties are present in if statements, return statements, loops, and function calls as well.

**Comment by onderaltintas** [comment]
Still there is no evidence that using "continue" is making code less testable and less maintainable for JavaScript. Saying that there are thousands of articles on the internet which say "continue is bad" is like saying that "snakes have legs" cause you have seen it on the internet from the newspaper named "Daily Testicle". I am re-closing it by applying the information there were no valid arguments on why "continue" should not be used.

**Comment by ricardodnb** [comment]
Sorry to continue commenting on this thread and this is opinionated but come on @ljharb, do you really prefer a bunch of possibly nested if-elses instead of simple continue statement? how can that be more readable ?
```javascript
for(const item of data) {
  if (x) {
    // ...
  } else {
    if (y) {
      // ...
    } else {
      // ...
    }
  }
}
```

vs

```javascript
for(const item of data) {
  if (x) {
    // ...
    continue
  }

  if (y) {
    // ...
    continue
  }

  // ...
}
```

**Comment by ljharb** [comment]
@ricardodnb yes. If you provide a non-contrived case, I’ll show you what i find most readable - and it’s not going to be a loop at all.

**Comment by icetbr** [comment]
@ricardodnb a concrete example would allow for better explanations, but  but your code could look something along these lines. 

```js
const itemsX = items.filter(isX).map(toItemX);
const itemsY = items.filter(isY).map(toItemY);
```

Avoiding loops means breaking down your operations into smaller steps. It makes your code more declarative then imperative.

**Comment by radarsu** [comment]
I'll add few points here for people that look for better justification of this rule:

1. Using continue, break as well as nested if statements or other complex logic inside loops suggests, that maybe we iterate over badly filtered list and we should split it into few lists, then run them through few different functions. This encourages writing smaller functions and smaller datasets (smaller pieces increase reusability) plus increases readability.
2. Unstructured control flow statements in general increase cognitive load when somebody reads the code. It's same for break, continue as well as multiple return statements from a single function, BUT pure functions are usually a chunk that is smaller, separated, testable and usually can be understood in reasonable amount of time. Also, when you understand separated function logic you can "compress" what it is doing in your brain and treat it as black box, allowing you to understand and fit into limited brain's capacity bigger picture of program works faster.
3. There are statements that are error-prone in programming languages. Goto, break, continue, labels, even very popular switch statement. They are error-prone for various reasons. Even if you can use them perfectly well and only in specific cases where they actually improve readability, there are majority of cases where you or somebody else could use them wrong and cause bugs. As programming is usually cooperative work, it's better if you enforce a rule to not use them at all.
4. Avoiding continue/break statement's sometimes enforces you to write code in more functional and immutable manner.

**Comment by bmind7** [comment]
All control flows are masqueraded `GOTO`s, even `IF` statement. Should we get rid of `IF`s? 

There are cases when we have to use `IF`, `CONTINUE` and so on to have better performance. And it doesn't mean you necessarily end up with less readable code. Running `FITLER` on the same array many times just to avoid `IF` and `CONTINUE` will kill performance and increase the noise to signal ratio of the codebase. 

Still, I believe there are many cases when we can use `FILTER`, `MAP`, `REDUCE` and have cleaner code, but there are no silver bullets.

**Comment by ljharb** [comment]
yes, it does mean that - performance overrides clarity here, and you’d use an eslint override comment to explain why you used a worse pattern.

**Comment by ljharb** [comment]
If you're considering performance at a given location, you're already thinking at a very, very low level.

**Comment by gonzalo-tajmuch-uno** [comment]
> [@ricardodnb](https://github.com/ricardodnb) a concrete example would allow for better explanations, but but your code could look something along these lines.
> 
> const itemsX = items.filter(isX).map(toItemX);
> const itemsY = items.filter(isY).map(toItemY);
> Avoiding loops means breaking down your operations into smaller steps. It makes your code more declarative than imperative.

Sorry to bump an old thread, but I'm sorry. Are you seriously implying you SHOULD create 2 datasets and then use foreach/map/etc in each dataset?

Do you realize a closure is created for each .filter iteration? and then for each .foreach iteration? Also creating more memory because now I'm creating two datasets and then iterating, instead of doing this in O(n) without allocating more memory than needed

The imperative vs functional debate focusing on what "looks nicer" gives me the impression nobody really cares about the performance of their code. Sure for a lot of use cases you can get away with things like .filter().map() in various places of your page. But arguing you SHOULD NOT use for loops, or that continue is harmful because people can use it wrong, is just mediocrity

**Comment by GonziTaji** [comment]
> [@ricardodnb](https://github.com/ricardodnb) a concrete example would allow for better explanations, but but your code could look something along these lines.
> 
> const itemsX = items.filter(isX).map(toItemX);
> const itemsY = items.filter(isY).map(toItemY);
> Avoiding loops means breaking down your operations into smaller steps. It makes your code more declarative than imperative.

Sorry to bump an old/closed issue, but I'm really concerned about this mentality and I feel it needs pushback.

You SHOULD avoid doing this unless you are working with small datasets or doing very trivial work, because a closure is created for every .filter iteration and then for every .foreach iteration (since "you should not use for loops in 2018"). Not to mention you are using unnecesary memory because now two arrays are being created just as intermediate values. And instead of all that, you can do everything in linear time, without allocating more memory than needed, and, if you know what you are doing, in a clear manner.

> If you're considering performance at a given location, you're already thinking at a very, very low level.

"Very, very low level", like caring about how performant is your webapp? Nice. This is why webapps usually take forever to load

Focusing on what "looks nicer" (The infamous imperative vs functional debate) gives me the impression nobody really cares about the performance of their code (backed by the above comment). Sure for a lot of use cases you can get away with things like doing .filter().map() multiple times in some cases. But arguing you SHOULD NOT use for loops is simply wrong. You SHOULD use what's NEEDED, not what a fad says what's more "nice looking", and disabling the es-lint rule is a bad practice in my opinion

Why? Es-lint rules, in my opinion, should be mandatory. Disabling es-lint rules every time you need to do something that someone argues looks ugly or could make an intern write bad code is a very bad practice, because then your team get used to disable es-lint rules any time they need to do anything non-trivial, and you lose control of which rules are actually important and should be enforced, and which rules can be disabled when needed. Any es-lint rule that is disabled inline more than once should not be an es-lint rule
